skywalking/agent/__init__.py

```diff
@@ -382,7 +414,272 @@ def notify_profile_finish(self, task: ProfileTask):
+            self.__log_queue = asyncio.Queue(maxsize=config.agent_log_reporter_max_buffer_size)
+        if config.agent_profile_active:
+            self.__snapshot_queue = asyncio.Queue(maxsize=config.agent_profile_snapshot_transport_buffer_size)
+        # initialize background coroutines
+        self.background_coroutines = set()
+        self.background_tasks = set()
+        self._finished = asyncio.Event()
+
+        # Install log reporter core
+        # TODO - Add support for printing traceID/ context in logs
+        if config.agent_log_reporter_active:
+            from skywalking import log
+            log.install()
+        if config.agent_meter_reporter_active:
+            # meter.init(force=True)
+            await meter.init_async()
+
+        self.__bootstrap()  # gather all coroutines
+
+        logger.debug('All background coroutines started')
+        await asyncio.gather(*self.background_coroutines)
+
+    def __start_event_loop(self) -> None:
+        try:
+            asyncio.run(self.__start_event_loop_async())
+        except asyncio.CancelledError:
+            logger.info('Python agent asyncio event loop is closed')
+        except Exception as e:
+            logger.error(f'Error in Python agent asyncio event loop: {e}')
+        finally:
+            if self._finished is not None:
+                self._finished.set()
+
+    def start(self) -> None:
+        loggings.init()
+
+        if sys.version_info < (3, 7):
+            # agent may or may not work for Python 3.6 and below
+            # since 3.6 is EOL, we will not officially support it
+            logger.warning('SkyWalking Python agent does not support Python 3.6 and below, '
+                           'please upgrade to Python 3.7 or above.')
+
+        if not self.__started:
+            # if not already started, start the agent
+            config.finalize()  # Must be finalized exactly once
+
+            self.__started = True
+            logger.info(f'SkyWalking async agent instance {config.agent_instance_name} starting in pid-{os.getpid()}.')
+
+            # Here we install all other lib plugins on first time start (parent process)
+            plugins.install()
+
+        elif self.__started and os.getpid() == self.started_pid:
+            # if already started, and this is the same process, raise an error
+            raise RuntimeError('SkyWalking Python agent has already been started in this process, '
+                               'did you call start more than once in your code + sw-python CLI? '
+                               'If you already use sw-python CLI, you should remove the manual start(), vice versa.')
+
+        self.started_pid = os.getpid()
+        atexit.register(self.__fini)
+
+        # still init profile here, since it is using threading rather than asyncio
+        if config.agent_profile_active:
+            profile.init()
+
+        # can use asyncio.to_thread() in Python 3.9+
```

Comment:
Is it any faster etc.? If so you could add a check to use to_thread. check if asyncio hasattr to_thread

Reply:
`asyncio.to_thread()` actually simply calls the pre-python 3.9 function `loop.run_in_executor()`, which uses `concurrent.futures.Executor` to create a thread pool to manage additional threads to execute specific coroutines. And here I simply use `Thread()` to create a new thread to execute the coroutine. Considering that we only need one extra thread to maintain the `eventloop`, the performance gap between the two should be negligible. In addition, `to_thread()`/`run_in_executor()` is an async function, which cannot actually be called here, and other codes need to be changed to achieve the same function as here. I write this comment mainly because I think `to_thread()` can make the code more concise and intuitive, but in fact it may not be necessary to use it at all.

I will consider removing this comment in future changes.